adrs/5541831634/README.md

```diff
@@ -0,0 +1,160 @@
+
+### Fractal pattern directories and files
+
+A fractal pattern is a pattern that you could repeat in each nested level
+directory. The pattern repeated in `<rootDir>/src`, `<rootDir>/src/routes/` or
+`<rootDir>/src/components/[component name]`.
+
+* You **must** use `<fractalDir>/routes` for managing the entry point of a
+  route.
+* You **must** use `<fractalDir>/services` to store anything
+  related to any communication with [external services](#external-services).
+* You **must** use external service brand name as a subdirectory in
+  `<fractalDir>/services/[service name]`.
+* You **must** use `<fractalDir>/components` to store reusable React
+  components. You may create a directory to scope the fractal directories to
+  a component, otherwise use use a file.
+* You **must** use `<fractalDir>/hooks` to store reusable React
+  hooks.
+* You **must** use `<fractalDir>/types` to store reusable TypeScript types.
+* You **must** use `<fractalDir>/styles` to store components styles such as
+  CSS or Stylesheets.
+* You **must** use `<fractalDir>/constants` to store constants variables.
+* You **must** use `<fractalDir>/contexts` to store React contexts definitions.
+* You **must** use `<fractalDir>/i18n` to store anything
+  related to Internationalization.
+* You **must** use `<fractalDir>/i18n/datetime` to store i18n date
+  and time formatting.
+* You **must** use `<fractalDir>/i18n/list` to store i18n sensitive
+  list formatting. Read more about the intent at [Intl.ListFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat).
+* You **must** use `<fractalDir>/i18n/numbers` to store i18n number
+  formatting. Read more about the intent at [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat).
+* You **must** use `<fractalDir>/i18n/plural` to store i18n
+  pluralization formatting. Read more about the intent at [Intl.PluralRules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules).
+* You **must** use `<fractalDir>/i18n/relative-time` to store i18n
+  relative-time formatting. Read more about the intent at [Intl.RelativeTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat).
+* You **must** use `<fractalDir>/helpers/datetime` to store anything related
+  to Dates or Times. Except formatters, use `<fractalDir>/i18n/datetime`
+  instead.
+* You **must** use `<fractalDir>/helpers/currency` to store anything related
+  to currency. Except formatters, use `<fractalDir>/i18n/numbers`
+  instead.
+* You **must** use `<fractalDir>/helpers` to store anything that does not
+  fall into any of the other helper directory categories. You **may** create
+  your own files and directories based on your context.
+* You **must** use `<fractalDir>/transports/http` to store anything
```

Comment:
More specifically I'm concerned that the structure may be _too_ granular in separating the different types of code.
It reminds me a lot of how we used to separate `style`, `js`, `html` in the early 2010s
```
styles/
  article.css
js/
  article.js
html/
  article.html
```

and then started colocating them
```
Article.jsx
```

If the intention of this ADR is to have an exercise on documenting your personal opinion on what a NextJS structure should look like, this is fine.

If the intention of this ADR is to have an exercise on:
- teaching others how to think about organizing code
- teach others why this structure is superior
- getting buy-in then I recommend breaking this ADR up into smaller chunks 

IE: do I18n, services, and transports as separate ADRs and provide simple but tangible examples

I say this because generally when defining a standard as large as this from within the bounds of an organization the burden of proof is on the person who wants to introduce the large standard. In this case you are the organization owner so you have the authority to implement without currying buy-in.

Reply:
> More specifically I'm concerned that the structure may be too granular in separating the different types of code. It reminds me a lot of how we used to separate style, js, html in the early 2010s

From my experience, people still living in 2010 then, all depends on the toolkit you are using 😉 

> And then started colocating them

If you want to colocate things into a single file, by all means (for example people in CSS-in-JS land), do that.

In case you need to find directories to group things, then you have an answer.

Style files are colocated together to allow easier collaboration with designers or have a precise directory where you can manage the styles (it is a back and forward, in 2010 Bootstrap was the way, in 2020 Tailwind 🤷🏻  don't get cut up in the example).

> I say this because generally when defining a standard as large as this from within the bounds of an organization the burden of proof is on the person who wants to introduce the large standard. In this case, you are the organization owner so you have the authority to implement without currying buy-in.

If you feel this is not the appropriate ADR to follow for your organization, then don't follow it. All ADRs come with a cost, and writers can't assess the entire internet that would like to adopt it since they don't have the scope of it.

Greenfields projects are always easier to introduce such ADRs, this ADR is not assuming that work needs to be done, but rather have a common ground for incoming projects, or have a path forward with legacy applications.

It is hard to tell the situation of a particular organization.

> IE: do I18n, services, and transports as separate ADRs and provide simple but tangible examples

We could add some examples of it, but, it depends. Would prefer to write another ADR for particular ways to do things.

> teach others why this structure is superior

It is not, trade-offs are required, what matters is the alignment rather than being superior, such things are hard to quantify and justify objectively.

Reply:
> Greenfields projects are always easier to introduce such ADRs, this ADR is not assuming that work needs to be done, but rather have a common ground for incoming projects, or have a path forward with legacy applications.

This is a helpful clarification.
I'd be happy to try this folder structure but it seems like the scope is so large that we'll inevitably need to make changes to it.

What's the process for amending ADRs?
Would we keep this in the `Trial` phase and then `Adopt` it after a few trials have occured?
Or would we go straight into `Adopt` and add additional ADRs to amend it?

Reply:
At the moment, you can replace an ADR as a way to amend it, if you will invalidate the ADR.

For the most part, it is a bad idea to update things but doesn't have to be like that, just that greater power comes with greater responsibilities.